pinot-core/src/main/java/com/linkedin/pinot/core/util/AvroUtils.java

```diff
@@ -160,21 +146,121 @@ public static Schema getPinotSchemaFromAvroSchemaFile(String avroSchemaFileName,
     return getPinotSchemaFromAvroSchema(avroSchema, fieldTypes, timeUnit);
   }
 
-  public static List<String> getAllColumnsInAvroFile(File avroFile) throws IOException {
-    final List<String> ret = new ArrayList<String>();
-    final DataFileStream<GenericRecord> reader = getAvroReader(avroFile);
-    for (final Field f : reader.getSchema().getFields()) {
-      ret.add(f.name());
+  /**
+   * Get the Avro file reader for the given file.
+   */
+  public static DataFileStream<GenericRecord> getAvroReader(File avroFile) throws IOException {
+    if (avroFile.getName().endsWith(".gz")) {
+      return new DataFileStream<>(new GZIPInputStream(new FileInputStream(avroFile)),
+          new GenericDatumReader<GenericRecord>());
+    } else {
+      return new DataFileStream<>(new FileInputStream(avroFile), new GenericDatumReader<GenericRecord>());
     }
-    reader.close();
-    return ret;
   }
 
-  public static DataFileStream<GenericRecord> getAvroReader(File avroFile) throws IOException {
-    if(avroFile.getName().endsWith("gz"))
-      return new DataFileStream<GenericRecord>(new GZIPInputStream(new FileInputStream(avroFile)), new GenericDatumReader<GenericRecord>());
-    else
-      return new DataFileStream<GenericRecord>(new FileInputStream(avroFile), new GenericDatumReader<GenericRecord>());
+  /**
+   * Return whether the Avro field is a single-value field.
+   */
+  public static boolean isSingleValueField(Field field) {
+    try {
+      org.apache.avro.Schema fieldSchema = extractSupportedSchema(field.schema());
+      return fieldSchema.getType() != org.apache.avro.Schema.Type.ARRAY;
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while extracting non-null schema from field: " + field.name(), e);
+    }
+  }
+
+  /**
+   * Extract the data type stored in Pinot for the given Avro field.
+   */
+  public static FieldSpec.DataType extractFieldDataType(Field field) {
+    try {
+      org.apache.avro.Schema fieldSchema = extractSupportedSchema(field.schema());
+      org.apache.avro.Schema.Type fieldType = fieldSchema.getType();
+      if (fieldType == org.apache.avro.Schema.Type.ARRAY) {
+        return FieldSpec.DataType.valueOf(extractSupportedSchema(fieldSchema.getElementType()).getType());
+      } else {
+        return FieldSpec.DataType.valueOf(fieldType);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while extracting data type from field: " + field.name(), e);
+    }
+  }
+
+  /**
+   * Helper method to extract the supported Avro schema from the given Avro field schema.
+   * <p>Currently we support INT/LONG/FLOAT/DOUBLE/BOOLEAN/STRING/ENUM
```

Comment:
How about Utf8 (as string)?

Reply:
Utf8 is not a type. For string type, Utf8 instance can be returned, we handled it in transformAvroValueToObject().